KAFKA-6241: Enable dynamic updates of broker SSL keystore

rajinisivaram opened the pull request:
Enable dynamic broker configuration (see KIP-226 for details). Includes
 - Base implementation to allow specific broker configs and custom configs to be dynamically updated
 - Extend DescribeConfigsRequest/Response to return all synonym configs and their sources in the order of precedence
 - Extend AdminClient to alter dynamic broker configs
 - Dynamic update of SSL keystores

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rajinisivaram committed:
KAFKA-6241: Enable dynamic updates of broker SSL keystore

tedyu commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -107,6 +133,48 @@ public String toString() {
     }
+
+
+    public enum ConfigSource {
```
Is this intended to be part of public API ?

rajinisivaram replied:
Yes, it is returned when describing configs.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -107,6 +133,48 @@ public String toString() {
+        DYNAMIC_BROKER_CONFIG,
+        DYNAMIC_DEFAULT_BROKER_CONFIG,
+        STATIC_BROKER_CONFIG,
+        DEFAULT_CONFIG
```
Maybe add comment specifying the difference from DYNAMIC_DEFAULT_BROKER_CONFIG

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -53,14 +56,31 @@ public ConfigEntry(String name, String value) {
+     * @param synonyms Synonym configs in order of precedence
+     */
+    ConfigEntry(String name, String value, ConfigSource source, boolean isSensitive, boolean isReadOnly,
+                       List<ConfigSynonym> synonyms) {
```
nit: weird alignment

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -53,14 +56,31 @@ public ConfigEntry(String name, String value) {
      * @param isDefault whether the config value is the default or if it's been explicitly set
      * @param isSensitive whether the config value is sensitive, the broker never returns the value if it is sensitive
      * @param isReadOnly whether the config is read-only and cannot be updated
+     * @deprecated
```
Maybe we should mention the version?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -99,14 +126,91 @@ public boolean isReadOnly() {
+     * Source of configuration entries.
+     */
+    public enum ConfigSource {
+        TOPIC_CONFIG,                   // dynamic topic config that is configured for a specific topic
```
Why don't we call this `DYNAMIC_TOPIC_CONFIG`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -99,14 +126,91 @@ public boolean isReadOnly() {
+    /**
+     * Class representing a configuration synonym of a {@link ConfigEntry}.
+     */
+
```
nit: unneeded newline

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1606,24 +1611,79 @@ private Resource configResourceToResource(ConfigResource configResource) {
+
+    private ConfigEntry.ConfigSource configSource(DescribeConfigsResponse.ConfigSource source) {
+        ConfigEntry.ConfigSource configSource;
+        switch (source) {
```
Maybe `DescribeConfigsResponse.ConfigSource` could have a method to get the corresponding `ConfigEntry.ConfigSource`?

rajinisivaram replied:
Don't know about that. `ConfigEntry.ConfigSource` is in the admin client's package, so not sure we want to refer to that from the `requests` package. At the moment, it uses the same pattern as `ConfigResource`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1606,24 +1611,79 @@ private Resource configResourceToResource(ConfigResource configResource) {
         }
+        NodeProvider nodeProvider;
+        ConfigResource resource = resources.iterator().next();
+        if (resource.type() == ConfigResource.Type.BROKER && !resource.name().isEmpty()) {
```
This redundant check is a tad annoying. Could we add the node supplier to the parameter list?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1494,7 +1494,7 @@ public DescribeConfigsResult describeConfigs(Collection<ConfigResource> configRe
 
         for (ConfigResource resource : configResources) {
-            if (resource.type() == ConfigResource.Type.BROKER) {
+            if (resource.type() == ConfigResource.Type.BROKER && !resource.name().isEmpty()) {
```
We check if the name is empty in a number of places. It might be clearer to turn it into a method and give it a good name.

rajinisivaram replied:
Don't know about `good` name, but have moved to a method and given it `a` name.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/Reconfigurable.java:
```diff
@@ -0,0 +1,47 @@
+     * all configs including any reconfigurable configs that may be different
+     * from the initial configuration.
+     */
+    boolean validate(Map<String, ?> configs);
```
I am wondering if we should name this `validateReconfiguration` to make it clear that it is not used for validating the initial configuration.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/Reconfigurable.java:
```diff
@@ -0,0 +1,47 @@
+    boolean validate(Map<String, ?> configs);
+
+    /**
+     * Reconfigures this instance with the given key-value pairs. The provided
```
Maybe emphasize that this will only be invoked if configs which have passed validation?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/ChannelBuilders.java:
```diff
@@ -70,18 +70,20 @@ public static ChannelBuilder clientChannelBuilder(SecurityProtocol securityProto
      * @return the configured `ChannelBuilder`
      */
     public static ChannelBuilder serverChannelBuilder(ListenerName listenerName,
+                                                      boolean interBrokerListener,
```
nit: `isInterBrokerListener`? Or maybe use a separate method `interBrokerChannelBuilder`?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/ChannelBuilder.java:
```diff
@@ -16,22 +16,24 @@
+     * is a client channel builder.
      */
-    void configure(Map<String, ?> configs) throws KafkaException;
+    ListenerName listenerName();
```
Maybe we should have a `ServerChannelBuilder`?

rajinisivaram replied:
We use a single channel builder for each type, so wasn't sure of introducing `ServerChannelBuilder`. Instead I moved `listenerName` out into another interface. Not all channel builders are reconfigurable, so the channel builders that do SSL now implement the new interface.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsRequest.java:
```diff
@@ -48,18 +50,29 @@
     private static final Schema DESCRIBE_CONFIGS_REQUEST_V0 = new Schema(
             new Field(RESOURCES_KEY_NAME, new ArrayOf(DESCRIBE_CONFIGS_REQUEST_RESOURCE_V0), "An array of config resources to be returned."));
 
+    private static final Schema DESCRIBE_CONFIGS_REQUEST_V1 = new Schema(
```
We should update `getErrorResponse` below for v1. Probably missing a test case.

rajinisivaram replied:
Oops, fixed and added test.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsResponse.java:
```diff
@@ -119,15 +153,65 @@ public boolean isSensitive() {
+        }
+
+        public static ConfigSource forId(byte id) {
+            for (ConfigSource source: values()) {
```
Could just do a `switch`?

rajinisivaram replied:
Switched to the same pattern as `ResourceType`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsResponse.java:
```diff
@@ -205,9 +322,23 @@ protected Struct toStruct(short version) {
                 configEntriesStruct.set(CONFIG_VALUE_KEY_NAME, configEntry.value);
                 configEntriesStruct.set(IS_SENSITIVE_KEY_NAME, configEntry.isSensitive);
-                configEntriesStruct.set(IS_DEFAULT_KEY_NAME, configEntry.isDefault);
+                if (configEntriesStruct.hasField(CONFIG_SOURCE_KEY_NAME))
```
There are `setIfExists` methods which work with  `Field` objects. We can try to use `Field` here or add a similar method to work with strings.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -64,8 +76,13 @@ public SslFactory(Mode mode) {
+        this(mode, clientAuthConfigOverride, false);
+    }
+
+    public SslFactory(Mode mode, String clientAuthConfigOverride, boolean clientAndServer) {
```
It's a little hard to guess what `clientAndServer` is doing from its name. I am guessing it means both client and server authentication?

rajinisivaram replied:
I meant that the configs can be used for both client-side and server-side, but agree that the name is confusing. Renamed to `keystoreVerifiableUsingTruststore` since that is what we are actually verifying.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -109,23 +126,66 @@ else if (clientAuthConfig.equals("requested"))
+            this.sslContext = newSslContext;
+    }
+
+    private SSLContext maybeCreateNewSSLContext(Map<String, ?> configs, boolean verifyKeystore, boolean validateOnly) throws KafkaException {
```
I'd suggest getting rid of the `validateOnly` parameter and creating two methods: one which validates only and one which validates and, if successful, updates the key store.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -181,38 +251,42 @@ public SSLContext sslContext() {
-        }
+            return new SecurityStore(type, path, password, keyPassword);
+        } else
+            return null;
```
It doesn't look like the code protects from a null key store.

rajinisivaram replied:
Keystore can be null for clients, added a comment.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
         }
     }
 
+    static class SSLConfigValidatorEngine {
```
Some high-level comments would be helpful to understand exactly what validation is being done.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+    }
+
+    static class CertificateEntries {
+        private final Principal subjectDn;
```
nit: maybe we may as well spell out "DistinguishedName."

rajinisivaram replied:
Changed to `subjectPrincipal`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+        }
+
+        CertificateEntries(X509Certificate cert) throws GeneralSecurityException {
+            this.subjectDn = cert.getSubjectDN();
```
In the documentation, `getSubjectDN` is marked as "denigrated" in favor of `getSubjectX500Principal`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+                            case OK: break;
+                            case BUFFER_OVERFLOW:
+                                netBuffer.compact();
+                                Utils.ensureCapacity(netBuffer, sslEngine.getSession().getPacketBufferSize());
```
Don't we need to use the buffer returned by `ensureCapacity`? A couple more below.

rajinisivaram replied:
Oops, fixed, thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+                                Utils.ensureCapacity(netBuffer, sslEngine.getSession().getPacketBufferSize());
+                                netBuffer.flip();
+                                break;
+                            case BUFFER_UNDERFLOW: throw new SSLException("Unexpected status: CLOSED");
```
Is "CLOSED" a typo here? Maybe we could fall through and use `handshakeStatus` in the message? Also, should we add a `default` case for cleanliness?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+                        }
+                        return;
+                    case NEED_UNWRAP:
+                        if (peerEngine.netBuffer.position() == 0)
```
Some of these checks are a little obscure and could use a brief comment.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+                    case FINISHED:
+                        return;
+                    case NOT_HANDSHAKING:
+                        if (handshakeResult.getHandshakeStatus() != SSLEngineResult.HandshakeStatus.FINISHED)
```
Can you explain why we need this check? Aren't we already checking the result of `getHandshakeStatus()`?

rajinisivaram replied:
I am not sure we hit this case, but I wanted to be cautious. It is because both handshake result and sslEngine store handshake status and they may not be the same.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+                        break;
+                    case NEED_TASK:
+                        sslEngine.getDelegatedTask().run();
+                        handshakeStatus = sslEngine.getHandshakeStatus();
```
As far as I can tell, every time we iterate the loop, we update `handshakeStatus`. Would it make sense to move its initialization to the first line in the loop and remove these updates?

rajinisivaram replied:
For the `unwrap` case, we get the result from the handshake result of unwrap, so it is different from delegated tasks where the latest status is from the sslEngine. The logic here is the same as `SslTransportLayer` and that contains a lot more comments. Unfortunately it would be too messy to combine the two, so I tried to make this as small as possible.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
 
+    static class SSLConfigValidatorEngine {
+        private static final ByteBuffer EMPTY_BUF = ByteBuffer.allocate(0);
+        private final SSLEngine sslEngine;
```
The `SSLEngine` documentation mentions that the engine should be closed after we have finished with it. It wasn't clear to me if that was needed here.

rajinisivaram replied:
Don't think it is needed here, but added anyway.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java:
```diff
@@ -76,4 +80,59 @@ public void testClientMode() throws Exception {
+                Mode.SERVER, trustStoreFile, "server");
+        SslFactory sslFactory = new SslFactory(Mode.SERVER);
+        sslFactory.configure(serverSslConfig);
+        sslFactory.createSSLContext(securityStore(serverSslConfig), true);
```
Maybe we could at least assert the resulting ssl context is not null?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+        private final Principal subjectDn;
+        private final Collection<List<?>> subjectAltNames;
+
+        static List<CertificateEntries> certificateEntries(KeyStore keystore) throws GeneralSecurityException, IOException {
```
This name is redundant in usage: `CertificateEntries.certificateEntries(...)`. Would `create` work?

hachikuji commented on clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java:
```diff
@@ -76,4 +80,59 @@ public void testClientMode() throws Exception {
     }
 
+    @Test
+    public void testKeyStoreTrustStoreValidation() throws Exception {
```
Should we have a basic sanity test case for reconfiguration itself? I looked for one and couldn't find any.

rajinisivaram replied:
Added in `SslTransportLayerTest`.

hachikuji reviewed (commented):
Thanks for the patch. I'm still working my way through the server-side code, but thought I may as well post my comments for the client-side stuff.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thank you for the review. I have addressed the comments so far.

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -280,24 +282,74 @@ class AdminManager(val config: KafkaConfig,
 
-      def createResponseConfig(config: AbstractConfig, isReadOnly: Boolean, isDefault: String => Boolean): DescribeConfigsResponse.Config = {
-        val filteredConfigPairs = config.values.asScala.filter { case (configName, _) =>
+      def configType(name: String, synonyms: List[String]): ConfigDef.Type = {
```
Quite a lot of inner defs for one method. Any harm pulling them up to the top level? I think it would improve readability of this function.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -53,14 +56,31 @@ public ConfigEntry(String name, String value) {
+     * @deprecated since 1.1.0. This constructor will be removed in a future release.
      */
     public ConfigEntry(String name, String value, boolean isDefault, boolean isSensitive, boolean isReadOnly) {
+        this(name, value, ConfigSource.UNKNOWN, isSensitive, isReadOnly, Collections.<ConfigSynonym>emptyList());
```
I'm not sure it actually matters since users are unlikely to construct this object manually, but it seems like we should use `ConfigSource.DEFAULT_CONFIG` if `isDefault` is true and `ConfigSource.UNKNOWN` otherwise. Then `isDefault()` will continue to work with this constructor.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -110,7 +146,6 @@ public boolean equals(Object o) {
 
         return this.name.equals(that.name) &&
                 this.value != null ? this.value.equals(that.value) : that.value == null &&
-                this.isDefault == that.isDefault &&
```
Should we include the config source instead for compatibility? Also, just double-checking that it is intentional to leave the synonyms out of `equals`.

rajinisivaram replied:
Added `source` and `synonyms`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1606,24 +1611,75 @@ private Resource configResourceToResource(ConfigResource configResource) {
     }
 
+    private List<ConfigEntry.ConfigSynonym> configSynonyms(DescribeConfigsResponse.ConfigEntry configEntry) {
+        List<ConfigEntry.ConfigSynonym> synonyms = new ArrayList<>();
```
nit: we may as well initialize the list with a size?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1606,24 +1611,75 @@ private Resource configResourceToResource(ConfigResource configResource) {
         final long now = time.milliseconds();
         runnable.call(new Call("alterConfigs", calcDeadlineMs(now, options.timeoutMs()),
-                new LeastLoadedNodeProvider()) {
+                nodeProvider) {
```
nit: move to previous line

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/ChannelBuilders.java:
```diff
@@ -95,13 +97,14 @@ private static ChannelBuilder create(SecurityProtocol securityProtocol,
                 JaasContext jaasContext = JaasContext.load(contextType, listenerName, configs);
-                channelBuilder = new SaslChannelBuilder(mode, jaasContext, securityProtocol, listenerName,
+                channelBuilder = new SaslChannelBuilder(mode, jaasContext, securityProtocol,
+                        listenerName, isInterBrokerListener,
```
nit: parameter list looks imbalanced

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsResponse.java:
```diff
@@ -155,9 +240,42 @@ public DescribeConfigsResponse(Struct struct) {
-                configEntries.add(new ConfigEntry(configName, configValue, isSensitive, isDefault, readOnly));
+                Collection<ConfigSynonym> synonyms;
+                if (configEntriesStruct.hasField(CONFIG_SYNONYMS_KEY_NAME)) {
+                    synonyms = new ArrayList<>();
```
nit: instantiate with size?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -148,10 +207,21 @@ private SSLContext createSSLContext() throws GeneralSecurityException, IOExcepti
+        if (verifyKeystore) {
+            if (keystoreVerifiableUsingTruststore)
+                SSLConfigValidatorEngine.validate(this, sslContext);
+            if (!CertificateEntries.create(this.keystore.load()).equals(CertificateEntries.create(keystore.load()))) {
```
Seems we assume that `verifyKeystore` will only be enabled if there is a non-null keystore. Maybe we can add a comment or move keystore verification to a separate method? Out of curiosity, why not always verify if the keystore is not null?

rajinisivaram replied:
Removed `verifyKeystore` flag

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -280,24 +282,74 @@ class AdminManager(val config: KafkaConfig,
+      }
+      def configSynonyms(name: String, synonyms: List[String], isSensitive: Boolean) : List[DescribeConfigsResponse.ConfigSynonym] = {
+        val dynamicConfig = config.dynamicConfig
+        val allSynonyms = synonyms.flatMap(name => configSynonym(name, dynamicConfig.dynamicBrokerConfigs,
```
nit: seems like a case where a mutable collection and a single pass over the synonym list might be a tad more natural (and less verbose).

rajinisivaram replied:
Changed to a mutable collection. But left it as multiple passes over the synonyms to preserve order.

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -367,11 +419,37 @@ class AdminManager(val config: KafkaConfig,
+
+            val configProps = new Properties
+            config.entries.asScala.foreach { configEntry =>
+              configProps.setProperty(configEntry.name(), configEntry.value())
```
nit: remove parenthesis for `name` and `value`

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -310,15 +362,15 @@ class AdminManager(val config: KafkaConfig,
             val topicProps = adminZkClient.fetchEntityConfig(ConfigType.Topic, topic)
             val logConfig = LogConfig.fromProps(KafkaServer.copyKafkaConfigToLog(config), topicProps)
-            createResponseConfig(logConfig, isReadOnly = false, name => !topicProps.containsKey(name))
+            createResponseConfig(logConfig, (name, value) => createTopicConfigEntry(name, value, topicProps))
```
A curried function would be slightly less verbose:
```scala
def createTopicConfigEntry(topicProps: Properties)(name: String, value: Any)
```

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -367,11 +419,37 @@ class AdminManager(val config: KafkaConfig,
             resource -> ApiError.NONE
+
+          case ResourceType.BROKER =>
+            val brokerId = if (resource.name == null || resource.name.isEmpty) None else try Some(resource.name.toInt) catch {
```
This is basically the same snippet we have above. Maybe we can move it to a method?

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -367,11 +419,37 @@ class AdminManager(val config: KafkaConfig,
+            }
+            resource -> ApiError.NONE
           case resourceType =>
             throw new InvalidRequestException(s"AlterConfigs is only supported for topics, but resource type is $resourceType")
```
This message needs to be updated.

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -367,11 +419,37 @@ class AdminManager(val config: KafkaConfig,
-        case e: ConfigException =>
+        case e @ (_: ConfigException | _: IllegalArgumentException) =>
           val message = s"Invalid config value for resource $resource: ${e.getMessage}"
           info(message)
```
Do you think it's intentional for the log level to be so high for an invalid request?

rajinisivaram replied:
It is like this in all the methods in AdminManager, but perhaps it shouldn't be? I will raise a JIRA to change it.

rajinisivaram replied:
Raised KAFKA-6463.

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -367,11 +419,37 @@ class AdminManager(val config: KafkaConfig,
+            }
+            val properties = this.config.dynamicConfig.validateAndConvertForPersistence(brokerId, configProps)
+
+            alterConfigPolicy match {
```
There's a bit of annoyance in the topic config case where we invoke `validateTopicConfig` and then `changeTopicConfig` if there is a policy set, but we skip the call to `validateTopicConfig` otherwise (since `changeTopicConfig` also invokes it). I feel this is kind of a silly optimization since config alterations are rare. Instead, we can change the logic to call `validateTopicConfig` prior to matching the policy. In that case, these snippets are almost identical as far as I can tell and we can factor the logic into a separate function.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+
+  // Map topic config to the broker config with highest priority. Some of these have additional synonyms
+  // that can be obtained using `brokerConfigSynonyms`
+  val topicConfigSynonyms = Map(
```
Would it make sense to move this into `LogConfig`? It may be useful to know this mapping generally and not just for dynamic broker configuration.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+      case KafkaConfig.LogRetentionTimeMillisProp | KafkaConfig.LogRetentionTimeMinutesProp | KafkaConfig.LogRetentionTimeHoursProp =>
+        List(KafkaConfig.LogRetentionTimeMillisProp, KafkaConfig.LogRetentionTimeMinutesProp, KafkaConfig.LogRetentionTimeHoursProp)
+      case listenerConfigRegex(baseName) if matchListenerOverride => List(name, baseName)
+      case n => List(n)
```
Or:
```scala
case _ => List(name)
```

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+    val adminZkClient = new AdminZkClient(zkClient)
+    updateDefaultConfig(adminZkClient.fetchEntityConfig(ConfigType.Broker, ConfigEntityName.Default))
+    updateBrokerConfig(brokerId, adminZkClient.fetchEntityConfig(ConfigType.Broker, brokerId.toString))
+    updateCurrentConfig()
```
Redundant since `updateBrokerConfig` calls this?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+    }
+  }
+
+  private def configToMap(config: Map[String, _]): immutable.Map[String, String] = {
```
Not sure if we could find a utility or config-related class where something general like this could live? Seems like we could just change `convertToString` to use the object type instead of `ConfigDef.Type`.

rajinisivaram replied:
I don't think we would want to use this anywhere else since it converts passwords to the actual value - we want to avoid doing this elsewhere. Have added a comment to clarify that.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+    toPersistentProps(props)
+  }
+
+  private def validateConfigTypes(brokerId: Option[Int], props: Properties, logError: Boolean): Unit = {
```
`brokerId` unused?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+    updateCurrentConfig()
+  }
+
+  private[server] def validateAndConvertForPersistence(brokerId: Option[Int], props: Properties): Properties = CoreUtils.inReadLock(lock) {
```
nit: since there is only one caller, it seems like we could just separate validation and conversion.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+    currentConfig = processConfig(newProps, validateOnly = false)
+  }
+
+  private def processConfig(newProps: mutable.Map[String, String], validateOnly: Boolean): KafkaConfig = {
```
nit: I don't think `newProps` needs to be mutable.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+  }
+
+  private def processConfig(newProps: mutable.Map[String, String], validateOnly: Boolean): KafkaConfig = {
+    val newConfig = new KafkaConfig(newProps.asJava, true, Some(this))
```
Do we want logging enabled if we are only validating? We skip error logging below if `validateOnly` is enabled.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+
+  private def overrideProps(props: mutable.Map[String, String], propsOverride: mutable.Map[String, String]): Unit = {
+    propsOverride.foreach { case (k, v) =>
+      brokerConfigSynonyms(k, matchListenerOverride = false).foreach(props.remove)
```
Brief comment why `matchListenerOverride` is false would be nice. I guess it is because the listener overrides have highest precedence?

rajinisivaram replied:
Yes, and also because the listener synonyms are the base listener configs which could be used by other listeners. Have added a comment.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,280 @@
+import scala.collection._
+import scala.collection.JavaConverters._
+
+object DynamicBrokerConfig {
```
Some high level comments would be helpful here, in particular in regard to config precedence.

hachikuji commented on core/src/main/scala/kafka/server/DynamicConfig.scala:
```diff
@@ -57,10 +59,12 @@ object DynamicConfig {
     def names = brokerConfigDef.names
 
-    def validate(props: Properties) = DynamicConfig.validate(brokerConfigDef, props)
+    def validate(props: Properties) = DynamicConfig.validate(brokerConfigDef, props, true, nonDynamicProps)
```
Can we name the boolean parameter in the calls to `validate`?

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1234,4 +1263,4 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean) extends Abstra
       s"${KafkaConfig.QueuedMaxBytesProp} must be larger or equal to ${KafkaConfig.SocketRequestMaxBytesProp}")
   }
-}
+}
```
nit: add back newline

hachikuji commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -0,0 +1,458 @@
+                                   consumer: KafkaConsumer[Array[Byte], Array[Byte]],
+                                   numRecords: Int,
+                                   topic: String = topic): Unit = {
+    val producerRecords = (1 to numRecords).map(i => new ProducerRecord(topic, s"key$i".getBytes,
```
Not a big deal, but if we want to use strings, we can use their corresponding serializers.

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thank you for the review. I have addressed the comments so far. I have a question about config precedence. I have done it one way and added comments in `DynamicBrokerConfig`. But I wasn't sure whether it is too confusing for users.

As an example, we have `log.roll.ms` and `log.roll.hours` where `log.roll.hours` is used if `log.roll.ms` is not specified. We allow dynamic update of all the topic-related configs. If `log.roll.ms` is defined in server.properties and you then do an update to set `log.roll.hours`, then we use the dynamically updated config from `log.roll.hours`. Do you think that is ok? Or should we only allow dynamic update of the highest priority config (`log.roll.ms`) in this case? I have no idea why we have all these variations. `AdminClient#describeConfigs` will return all these synonyms in the order of precedence, so perhaps it is ok, but wanted a second opinion. Thank you!

rajinisivaram committed:
Improve error handling, cache current config in KafkaConfig, unit test

rajinisivaram commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -132,9 +175,76 @@ public String toString() {
+    /**
+     * Class representing a configuration synonym of a {@link ConfigEntry}.
+     */
+    public static class ConfigSynonym {
```
I think you need `equals/hashCode` since `synonyms` is now included in the same methods for `ConfigEntry`.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -237,10 +237,14 @@ public void testSerialization() throws Exception {
-        checkRequest(createDescribeConfigsRequest());
-        checkRequest(createDescribeConfigsRequestWithConfigEntries());
-        checkErrorResponse(createDescribeConfigsRequest(), new UnknownServerException());
+        checkRequest(createDescribeConfigsRequest((short) 0));
```
nit: we have tended to use an `int` for the version in other cases to avoid the annoying typecasts in the caller.

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -392,4 +402,77 @@ class AdminManager(val config: KafkaConfig,
+    val dynamicConfig = config.dynamicConfig
+    val allSynonyms = mutable.Buffer[DescribeConfigsResponse.ConfigSynonym]()
+
+    def maybeAddSynonym(name: String, map: Map[String, String], isSensitive: Boolean, source: ConfigSource): Unit = {
```
nit: do we need `isSensitive` in this argument list since we are not overriding the passed value?

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -392,4 +402,77 @@ class AdminManager(val config: KafkaConfig,
+      }
+    }
+
+    synonyms.foreach(name => maybeAddSynonym(name, dynamicConfig.currentDynamicBrokerConfigs,
```
My suggestion may not have been very clear before, but basically we can define `maybeAddSynonym` like this:

```scala
def maybeAddSynonym(map: Map[String, String], source: ConfigSource)(name: String)
```

Then this simplifies to this:
```scala
synonyms.foreach(maybeAddSynonym(dynamicConfig.currentDynamicBrokerConfigs, ConfigSource.DYNAMIC_BROKER_CONFIG)
```

hachikuji commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -392,4 +402,77 @@ class AdminManager(val config: KafkaConfig,
+        new DescribeConfigsResponse.ConfigSynonym(name, valueAsString, ConfigSource.TOPIC_CONFIG) +: list
+    }
+    val source = if (allSynonyms.isEmpty) ConfigSource.DEFAULT_CONFIG else allSynonyms.head.source
+    val synonyms = if (!includeSynonyms) List.empty else allSynonyms
```
It's a tad vexing that we build the full synonym list only to pick the first entry when `includeSynonyms` is false. Probably not too big of a deal though.

rajinisivaram replied:
Since it is not a performance-critical code path, it felt easier to use a single method.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,378 @@
+  * some configs have additional synonyms.
+  * </p>
+  * <ul>
+  *   <li>Listener configs may be defined using the prefix <tt>listener.name.listenerName.configName</tt>. These may be
```
Maybe this could be written as `listener.name.{listenerName}.{configName}` so that the variable parts are clear.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,378 @@
+  private[server] def validate(props: Properties, perBrokerConfig: Boolean): Unit = CoreUtils.inReadLock(lock) {
+    def checkInvalidProps(invalidPropNames: Set[String], errorMessage: String): Unit = {
+      if (invalidPropNames.nonEmpty) {
+        invalidPropNames.foreach(props.remove)
```
Why do we remove from `props`?

rajinisivaram replied:
Copy-paste error, removed.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,378 @@
+    }
+    checkInvalidProps(nonDynamicConfigs(props), "Cannot update these configs dynamically")
+    checkInvalidProps(securityConfigsWithoutListenerPrefix(props),
+      "Security configs can be dynamically updated only per-listener using the listener prefix")
```
nit: `These security configs...`?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,378 @@
+  * </p>
+  * <ul>
+  *   <li>Listener configs may be defined using the prefix <tt>listener.name.listenerName.configName</tt>. These may be
+  *       configured as dynamic or static broker configs. Listener configs have higher precedence than the base configs
```
To clarify my understanding, you must override listener configs at the broker level? So if all brokers share the same listener, and you want to update a listener property, you have to do it for every broker?

rajinisivaram replied:
Yes, in the current implementation, listener configs can be configured only per-broker. Some listener configs have to be defined at the broker level anyway (e.g. keystore/keystore password, Kerberos JAAS config that contains hostname). And the actual listeners/advertised listeners also need to be per-broker since they may contain host. At least initially, I thought it would best to keep all these related configs together so that they get reconfigured at the same time. Keystores are the only configs for which we allow updates in this KIP and these need to be per-broker. For other configs, we only support addition of new listeners, so configuring per-broker with all the listener configs should be ok?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,378 @@
+    * actual password values to String. Values from this map should never be used in log entries.
+    */
+  private def configToMap(config: Map[String, _]): immutable.Map[String, String] = {
+    config.filter(_._2 != null).map {
```
I still think it would be reasonable to move the conversion logic over to `ConfigDef`

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -0,0 +1,378 @@
+    }
+  }
+
+  private def processConfig(newProps: Map[String, String], validateOnly: Boolean): KafkaConfig = {
```
Maybe `processReconfiguration`?

hachikuji reviewed (commented):
Left a few more comments. Looking good overall.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thank you for the review. I have addressed the comments and rebased.

hachikuji commented:
retest this please

hachikuji reviewed (approved):
@rajinisivaram LGTM. Thanks for the patch! I haven't seen the failures in `testProducerConsumerOverrideUnthrottled` before. Can you confirm that they are not related?

rajinisivaram commented:
@hachikuji Thank you for the review. There are some transient quota test failures, but I am not sure if this particular test has an issue. I will run the build again and also run it locally a few times to check.

rajinisivaram commented:
retest this please

rajinisivaram commented:
retest this please

allenxiang commented:
@rajinisivaram @hachikuji  Is there any plan to enable dynamic updates of broker SSL truststore? Also it would be very useful to check if keystore/truststore's last modified time and reload it automatically if needed without user intervention.

rajinisivaram commented:
@allenxiang Only keystore update has been implemented for the 1.1 release. We started with a subset of configs that tend to get updated frequently. But now that the framework for dynamic updates is there, we can make more configs dynamic. Dynamic update of truststores should be relatively straightforward. If we enable truststore update using AdminClient/kafka-configs.sh, would you still require automatic reload?

allenxiang commented:
@rajinisivaram For our use case (DataHub at Monsanto), automatic reload is way more useful than using AdminClient/kafka-configs.sh since it seems kafka-configs.sh requires new file path/password etc to trigger reload. We don't change file path/password often but change the entries in the truststore quite often. Thanks.

rajinisivaram commented:
@allenxiang At the moment, for keystore updates, you do have to change file path and/or password. But we could in theory we could allow updates without that change. Would that be sufficient to address your use case?

allenxiang commented:
@rajinisivaram What we do right now is having a separate micro service in charge of updating the truststore file content and have our patched kafka brokers pick up changes automatically based on the last modified timestamp. No human intervention is needed. But if you guys implement the feature to allow us reloading truststore using AdminClient/kafka-configs.sh without requiring file path or password change, I should be able to call the functions used in kafka-configs.sh from our micro service to request reloading of the truststore. 

I have created a new pull request based on the current trunk for your reference on how we patch kafka. https://github.com/apache/kafka/pull/4541. I still hope you guys can incorporate this change since it is a small change but quite useful in term of minimize operation efforts. Thanks.

jcalcote commented:
I know I'm late on this thread, but I agree with @allenxiang that it's more useful to dynamically reload the store based on content changes than on configuration changes referring to a different file or password or store type. Furthermore, reloading of a trust store is more relevant than that of a key store, as new clients require client certs be added to the broker trust store.
